UPGRADING.md

```diff
@@ -0,0 +1,207 @@
+# Upgrading
+
+## Scenario 6.x to Scenario 7.x
+
+Scenario 7.0 had substantial API incompatibility with earlier versions, but
+comes with an intention to reduce the frequency of breaking changes in the
+future, alignin with the `ops` library.
+
+The changes listed below are not the only features introduced in Scenario 7.0
+(for that, see the release notes), but cover the breaking changes where you will
+need to update your existing Scenario tests.
+
+### Specifying events
+
+In previous versions of Scenario, an event would be passed to `Context.run`
+as a string name, via a convenient shorthand property of a state component
+(e.g. `Relation`, `Container`), or by explicitly constructing an `Event` object.
+These have been unified into a single `Context.on.{event name}()` approach,
+which is more consistent, resembles the structure you're familiar with from
+charm `observe` calls, and should provide more context to IDE and linting tools.
+
+```python
+# Older Scenario code.
+ctx.run('start', state)
+ctx.run(container.pebble_ready_event, state)
+ctx.run(Event('relation-joined', relation=relation), state)
+
+# Scenario 7.x
+ctx.run(ctx.on.start(), state)
+ctx.run(ctx.on.pebble_ready(container=container), state)
+ctx.run(ctx.on.relation_joined(relation=relation), state)
+```
+
+### Running custom events
+
+Scenario no longer supports explicitly running custom events. Instead, you
+should run the Juju event(s) that will trigger the custom event. For example,
+if you have a charm lib that will emit a `database-created` event on
+`relation-created`:
+
+```python
+# Older Scenario code.
+ctx.run("my_charm_lib.on.database_created", state)
+
+# Scenario 7.x
+ctx.run(ctx.on.relation_created(relation=relation), state)
+```
+
+Scenario will still capture custom events in `Context.emitted_events`.
+
+### State components are (frozen) sets, rather than lists
+
+State components, like containers, relations, and networks, do not have any
+inherant ordering. When these were lists, 'magic' numbers tended to creep into
+test code. These are now all sets, and have 'get' methods to retrieve the
+object you want to assert on. In addition, they are actually `frozenset`s
+(Scenario will automatically freeze them if you pass a `set`), which increases
+the immutability of the state and prevents accidentally modifying the input
+state.
+
+```python
+# Older Scenario code.
+state_in = State(containers=[c1, c2], relations=[r1, r2])
+...
+assert state_out.containers[1]...
+assert state_out.relations[0]...
+state_out.relations.append(r3)  # Not recommended!
+
+# Scenario 7.x
+state_in = State(containers={c1, c2}, relations={r1, r2})
+...
+assert state_out.get_container(c2.name)...
+assert state_out.get_relation(id=r1.id)...
```

Comment:
Is the keyword `id=` necessary here? You don't use it for `get_container`, so seems odd to have it for `get_relation`.

Reply:
Annoyingly, yes, because you can get a relation by `id` or by `label`, and because it's not obvious which one a positional argument would refer to, it has to be by keyword. For containers, you can only use the name, so forcing `get_container(name=x)` seemed annoying.

(For reference, see the PR for adding keyword args, where there was lots of back and forth - you probably remember some of it - on what the signatures for these get methods should be).

Reply:
Great, thanks. Yeah, sorry, was too lazy to look up that discussion. :-)